File: src/no/omb/bigboat/BigBoat.java
package no.omb.bigboat;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import no.omb.bigboat.data.Boat;
import no.omb.bigboat.data.ClubEntry;
import no.omb.bigboat.data.RaceData;
import no.omb.bigboat.data.RaceEntry;
import no.omb.bigboat.data.SeriesEntry;
import no.omb.bigboat.io.ConsoleWriter;
import no.omb.bigboat.io.CsvReader;
import no.omb.bigboat.io.CsvWriter;
import no.omb.bigboat.io.DataReader;
import no.omb.bigboat.io.HtmlReader;
import no.omb.bigboat.io.HtmlWriter;

public class BigBoat {
	
	public static final String YEAR = "2015";
	public static final String DATA = "data-" + YEAR;
	public static final String RESULTS = "results-" + YEAR;
	public static final String CHARSET = "ISO-8859-1";
	public static final char SEP = ';';
	public static final int CLUB_MAX_SCORES = 5;

	private static Map<RaceData, List<RaceEntry>> raceResults = new HashMap<>();
	private static Map<Boat, SeriesEntry> seriesEntries = new HashMap<>();
	private static Map<String, ClubEntry> clubEntries = new HashMap<>();

	public static void main(String[] args) {
		for (RaceData race : RaceData.races) {
			List<RaceEntry> raceEntries = getDataReader(race).parseDataFile(race);
			if (raceEntries != null) {
				raceResults.put(race, raceEntries);
			}
		}
		buildSeries();
		logResults();
		writeResults();
	}

	private static void buildSeries() {
		for (RaceData race : RaceData.races) {
			for (RaceEntry entry : raceResults.get(race)) {
				Boat boat = entry.getBoat();
				SeriesEntry seriesEntry = seriesEntries.get(boat);
				if (seriesEntry == null) {
					seriesEntry = new SeriesEntry(boat);
					seriesEntries.put(boat, seriesEntry);
				}
				seriesEntry.getRaceEntries().put(race, entry);
			}
		}
		for (SeriesEntry seriesEntry : seriesEntries.values()) {
			seriesEntry.computeSeriesScore();
		}
		computeClubScores();
	}

	private static void computeClubScores() {
		List<SeriesEntry> list = getSortedSeries();
		for (SeriesEntry seriesEntry : list) {			
			String club = seriesEntry.getBoat().getClub();
			ClubEntry clubEntry = clubEntries.get(club);
			if (clubEntry == null) {
				clubEntry = new ClubEntry(club);
				clubEntries.put(club, clubEntry);
			}
			if (clubEntry.getNumScores() < CLUB_MAX_SCORES) {
				clubEntry.setScore(clubEntry.getScore() + seriesEntry.getScore());
				clubEntry.setNumScores(clubEntry.getNumScores() + 1);
			}
		}
		for (ClubEntry clubEntry : clubEntries.values()) {
			if (clubEntry.getNumScores() < CLUB_MAX_SCORES) {
				clubEntry.setScore(clubEntry.getScore() + ((CLUB_MAX_SCORES-clubEntry.getNumScores())*SeriesEntry.DEFAULT_SCORE*RaceData.races.length));
			}
		}
	}

	public static Map<RaceData, List<RaceEntry>> getRaceResults() {
		return raceResults;
	}

	public static List<SeriesEntry> getSortedSeries() {
		List<SeriesEntry> list = new ArrayList<>();
		list.addAll(seriesEntries.values());
		Collections.sort(list);
		return list;
	}

	public static List<ClubEntry> getSortedClubs() {
		List<ClubEntry> list = new ArrayList<>();
		list.addAll(clubEntries.values());
		Collections.sort(list);
		return list;
	}

	private static void logResults() {
		ConsoleWriter.getInstance().logRaces();
		ConsoleWriter.getInstance().logSeries();
		ConsoleWriter.getInstance().logClubs();
		ConsoleWriter.getInstance().logDuplicates();
	}

	private static void writeResults() {
		for (RaceData race : RaceData.races) {
			CsvWriter.getInstance().writeRaceResultFile(race);
		}
		HtmlWriter.getInstance().writeSeriesResultHtml();
		CsvWriter.getInstance().writeSeriesResultFile();
		CsvWriter.getInstance().writeClubsResultFile();
	}

	private static DataReader getDataReader(RaceData race) {
		switch (race.getDataFileType()) {
		case SAILWAVE_RACE_HTML:
		case SAILWAVE_SUMMARY_HTML:
			return HtmlReader.getInstance();
		case SEILMAG_CSV:
			return CsvReader.getInstance();
		}
		return null;
	}

}

File: src/no/omb/bigboat/BigBoat.kt
package no.omb.bigboat

import no.omb.bigboat.data.*
import no.omb.bigboat.data.RaceData.DataFileType
import no.omb.bigboat.io.*
import java.io.File
import java.util.*


object BigBoat {
    const val YEAR = "2023"
    const val DATA = "data-$YEAR"
    const val RESULTS = "results-$YEAR"
    const val CHARSET = "UTF-8"
    const val SEP = ';'
    const val CLUB_MAX_SCORES = 5
    const val CANCELS = 2
    private val raceResults: MutableMap<RaceData, List<RaceEntry>> = HashMap()
    private val seriesEntries: MutableMap<Boat, SeriesEntry> = HashMap()
    private val clubEntries: MutableMap<String, ClubEntry> = HashMap()

    @JvmStatic
    fun main(args: Array<String>) {
        for (race in RaceData.races) {
            val raceEntries = getDataReader(race).parseDataFile(race)
            raceResults[race] = raceEntries
        }
        buildSeries()
        logResults()
        writeResults()
    }

    private fun buildSeries() {
        for (race in RaceData.races) {
            for (entry in raceResults[race]!!) {
                val boat = entry.boat!!
                var seriesEntry = seriesEntries[boat]
                if (seriesEntry == null) {
                    seriesEntry = SeriesEntry(boat)
                    seriesEntries[boat] = seriesEntry
                }
                seriesEntry.raceEntries[race] = entry
            }
        }
        for (seriesEntry in seriesEntries.values) {
            seriesEntry.computeSeriesScore()
        }
        computeClubScores()
    }

    private fun computeClubScores() {
        val list = sortedSeries
        for (seriesEntry in list) {
            val club = seriesEntry.boat.club
            var clubEntry = clubEntries[club]
            if (clubEntry == null) {
                clubEntry = ClubEntry(club)
                clubEntries[club] = clubEntry
            }
            if (clubEntry.numScores < CLUB_MAX_SCORES) {
                clubEntry.score = clubEntry.score + seriesEntry.score
                clubEntry.numScores = clubEntry.numScores + 1
            }
        }
        for (clubEntry in clubEntries.values) {
            if (clubEntry.numScores < CLUB_MAX_SCORES) {
                clubEntry.score =
                    clubEntry.score + (CLUB_MAX_SCORES - clubEntry.numScores) * SeriesEntry.DEFAULT_SCORE * RaceData.races.size
            }
        }
    }

    fun getRaceResults(): Map<RaceData, List<RaceEntry>> {
        return raceResults
    }

    val sortedSeries: List<SeriesEntry>
        get() {
            val list: MutableList<SeriesEntry> = ArrayList()
            list.addAll(seriesEntries.values)
            list.sort()
            return list
        }
    val sortedClubs: List<ClubEntry>
        get() {
            val list: MutableList<ClubEntry> = ArrayList()
            list.addAll(clubEntries.values)
            list.sort()
            return list
        }

    private fun logResults() {
        ConsoleWriter.instance.logRaces()
        ConsoleWriter.instance.logSeries()
        ConsoleWriter.instance.logClubs()
        ConsoleWriter.instance.logDuplicates()
    }

    private fun writeResults() {
        File(RESULTS).mkdir()
        for (race in RaceData.races) {
            CsvWriter.instance.writeRaceResultFile(race)
        }
        HtmlWriter.instance.writeSeriesResultHtml()
        CsvWriter.instance.writeSeriesResultFile()
        CsvWriter.instance.writeClubsResultFile()
    }

    private fun getDataReader(race: RaceData): DataReader {
        return when (race.dataFileType) {
            DataFileType.SAILWAVE_RACE_HTML, DataFileType.SAILWAVE_SUMMARY_HTML, DataFileType.SAILWAVE_SUMMARY2_HTML -> HtmlReader.instance
            DataFileType.SEILMAG_CSV, DataFileType.SEILMAG_CSV2, DataFileType.SEILMAG_CSV3, DataFileType.SEILMAG_CSV4, DataFileType.ULLERN_CSV, DataFileType.M2S_CSV, DataFileType.M2S_CSV2, DataFileType.M2S_CSV3 -> CsvReader.instance
        }
    }
}
